assembly/src/ast/module.rs

```diff
@@ -1,321 +1,688 @@
-        let docs_len = source.read_u16()? as usize;
-        let docs = if docs_len != 0 {
-            let str = source.read_vec(docs_len)?;
-            let str =
-                from_utf8(&str).map_err(|e| DeserializationError::InvalidValue(e.to_string()))?;
-            Some(str.to_string())
-        } else {
-            None
-        };
+    /// Same as [Module::resolve_import], but returns a mutable reference to the [Import]
+    pub fn resolve_import_mut(&mut self, module_name: &Ident) -> Option<&mut Import> {
+        self.imports.iter_mut().find(|import| &import.name == module_name)
+    }
 
-        // deserialize imports if required
-        let import_info = if options.serialize_imports {
-            ModuleImports::read_from(source)?
-        } else {
-            ModuleImports::default()
-        };
+    /// Return an iterator over the paths of all imports in this module
+    pub fn import_paths(&self) -> impl Iterator<Item = &LibraryPath> + '_ {
+        self.imports.iter().map(|import| &import.path)
+    }
+}
+
+impl core::ops::Index<ProcedureIndex> for Module {
+    type Output = Export;
+
+    #[inline]
+    fn index(&self, index: ProcedureIndex) -> &Self::Output {
+        &self.procedures[index.as_usize()]
+    }
+}
+
+impl core::ops::IndexMut<ProcedureIndex> for Module {
+    #[inline]
+    fn index_mut(&mut self, index: ProcedureIndex) -> &mut Self::Output {
+        &mut self.procedures[index.as_usize()]
+    }
+}
 
-        // deserialize re-exports
-        let num_reexported_procs = source.read_u16()? as usize;
-        let reexported_procs = source.read_many::<ProcReExport>(num_reexported_procs)?;
+impl Eq for Module {}
 
-        // deserialize local procs
-        let num_local_procs = source.read_u16()? as usize;
-        let local_procs = source.read_many::<ProcedureAst>(num_local_procs)?;
+impl PartialEq for Module {
+    fn eq(&self, other: &Self) -> bool {
+        self.kind == other.kind
+            && self.path == other.path
+            && self.docs == other.docs
+            && self.imports == other.imports
+            && self.procedures == other.procedures
+    }
+}
 
-        match Self::new(local_procs, reexported_procs, docs) {
-            Err(err) => Err(DeserializationError::UnknownError(err.message().clone())),
-            Ok(res) => Ok(res.with_import_info(import_info)),
+/// Serialization
+impl Module {
+    /// Serialization this module to `target`, using `options`.
+    pub fn write_into_with_options<W: ByteWriter>(&self, target: &mut W, options: AstSerdeOptions) {
+        options.write_into(target);
+        if options.debug_info {
+            self.span.write_into(target);
+            if let Some(source_file) = self.source_file.as_ref() {
+                target.write_u8(1);
+                let source_name = source_file.name();
+                let source_bytes = source_file.inner().as_bytes();
+                target.write_usize(source_name.as_bytes().len());
```

Comment:
That will certainly help with deserialization, unfortunately we'd need an impl for `str` for serialization, since the underlying type here is an `Arc<str>`, not a `String`. I am hoping we can just get rid of this serialization code entirely soon though.

Reply:
Isn't it `name` that converts a [`String` to `&str`](https://github.com/0xPolygonMiden/miden-vm/blob/e7b623ebc60c3444d2187cfe85c4162610c07f8e/assembly/src/ast/module.rs#L275-L279)? [`SourceFile`](https://github.com/0xPolygonMiden/miden-vm/blob/e7b623ebc60c3444d2187cfe85c4162610c07f8e/assembly/src/diagnostics.rs#L9) seems like a `String`

Reply:
Ah, you're right, `inner()` here returns a reference to a `String`, I was thinking of the identifier types which are all `Arc<str>`

Reply:
Turns out that the version of `winterfell` in use here doesn't implement `Serializable` for `String`, but I know it's implemented upstream. When I update `winterfell` to pull in the changes from facebook/winterfell#262, I'll update this.

Reply:
Winterfell v0.8.3 crates with this change (and also with updates to the serialization code) are now on crates.io.

Reply:
Thanks @bobbinth! I'll update the PR with that change here in a bit.